taichi/codegen/codegen_opengl.cpp

```diff
@@ -849,14 +856,22 @@ void OpenglCodeGen::lower() {  // {{{
 #endif
 }  // }}}
 
+FunctionType make_launch_function(std::unique_ptr<CompiledProgram> p, GLSLLauncher *launcher)
+{
+  static std::vector<std::unique_ptr<CompiledProgram>> no_gc;
```

Comment:
Great, now that you have `launcher`, could you also move this to `launcher`? Moreover, just remove `make_launch_function` and inline the procedures in `gen`?

Reply:
You mean, `launcher->no_gc`? That requires a forward decl of `CompiledProgram` in `opengl_kernel_launcher.h`...

Reply:
I believe you can declare a member function on `GLSLLauncher`, something like `register_program(std::unique_ptr< CompiledProgram> prog)`? The implementation is put into `.cpp` file.

Also, i don't think you need to include `"opengl_kernel_launcher.h"` in `opengl_api.h`? Instead, you can include it in `opengl_api.cpp`? This might reduce some circular includes, if that's what you are worried about.